src/inject/dynamic-theme/variables.ts

```diff
@@ -0,0 +1,107 @@
+import type {Theme} from '../../definitions';
+import {modifyBackgroundColor, modifyBorderColor, modifyForegroundColor} from '../../generators/modify-colors';
+import type {RGBA} from '../../utils/color';
+import {parse, rgbToString} from '../../utils/color';
+import {logWarn} from '../utils/log';
+import {replaceCSSVariables} from './css-rules';
+
+export const legacyVariables = new Map<string, string>();
+export const variables = new Map<string, Variable>();
+export const cachedVariables = new Map<string, CachedVariables>();
+
+interface CachedVariables {
+    modifiedBackground: string;
+    modifiedText: string;
+    modifiedBorder: string;
+}
+
+interface VariableDeclaration extends CachedVariables {
+    selector: string;
+    parentGroups: string[];
+    key: string;
+    property: string;
+}
+
+export interface Variable {
+    property: string;
+    value: string;
+    parentGroups: string[];
+}
+
+export function updateVariables(newVars: Map<string, Variable>, theme: Theme) {
+    if (newVars.size === 0) {
+        return;
+    }
+    newVars.forEach((value, key) => {
+        legacyVariables.set(value.property, value.value);
+        cachedVariables.delete(key);
+        variables.set(key, value);
+    });
+
+    legacyVariables.forEach((value, key) => {
```

Comment:
Why there are these `legacyVariables` and why they are passed into Style Managers?

Reply:
So, because in plain CSS any variable e.g. `var(--text)` is unknown what the value would be, that's why I preserve the old implementation of replacing the variable which is value. Because we switches it later back to the original value it's pure for dark reader to know what the `value type` is e.g. an <color> or an <opacity>. An based on that won't style an opacity, but only an color variable. It still has the flaw of overwritten by a new selector, but the value type won't change most likely and shouldn't be a problem.